jq/exec.go

```diff
@@ -0,0 +1,314 @@
+		for iter := C.jv_object_iter(jv); C.jv_object_iter_valid(jv, iter) == 1; iter = C.jv_object_iter_next(jv, iter) {
+			jvKey := C.jv_object_iter_key(jv, iter)
+			jvValue := C.jv_object_iter_value(jv, iter)
+			// This is safe because jv_object_iter_key asserts the jv is a string.
+			key := C.GoString(C.jv_string_value(jvKey))
+
+			obj[key] = jvToGoValue(jvValue)
+
+			C.jv_free(jvKey)
+			C.jv_free(jvValue)
+		}
+		return obj
+	}
+	panic("unreachable")
+}
+
+func dumpJvToGoStr(jv C.jv) string {
+	// jv_dump_string frees the provided jv, so we copy it.
+	dumpedjv := C.jv_dump_string(C.jv_copy(jv), C.int(0))
+	defer C.jv_free(dumpedjv)
+
+	// Strings from jv_string_value are cleaned up when the jv is freed.
+	return C.GoString(C.jv_string_value(dumpedjv))
+}
+
+func jvInterface(i interface{}) C.jv {
+	if i == nil {
+		return C.jv_null()
+	}
+
+	// Handle simple types.
+	switch val := i.(type) {
+	case int, int8, int16, int32, int64,
+		uint, uint8, uint16, uint32, uint64,
+		float32, float64:
+		return jvNumber(val)
+	case string:
+		return jvString(val)
+	case []byte:
+		return jvString(string(val))
+	case bool:
+		if val {
+			return C.jv_true()
+		}
+		return C.jv_false()
+	}
+
+	// Handle complex types.
+	val := reflect.ValueOf(i)
+	switch val.Kind() {
+	case reflect.Array, reflect.Slice:
+		return jvArray(val)
+	case reflect.Map:
+		return jvMap(val)
+	}
+
+	panic("unknown type attempted to be cast to jv")
+}
+
+func jvMap(val reflect.Value) C.jv {
+	jvObj := C.jv_object()
+
+	for _, key := range val.MapKeys() {
+		// These allocations are freed when the whole object is freed.
+		keyJv := jvString(key.String())
+		valJv := jvInterface(val.MapIndex(key).Interface())
+		C.jv_object_set(jvObj, keyJv, valJv)
+	}
+
+	return jvObj
+}
+
+func jvArray(val reflect.Value) C.jv {
+	len := val.Len()
+	jvArray := C.jv_array_sized(C.int(len))
+
+	for i := 0; i < len; i++ {
+		// These allocations are freed when the whole array is freed.
+		jvArray = C.jv_array_set(jvArray, C.int(i), jvInterface(val.Index(i).Interface()))
+	}
+
+	return jvArray
+}
+
+func jvString(i interface{}) C.jv {
+	str := i.(string)
+	cstr := C.CString(str)
+	defer C.free(unsafe.Pointer(cstr))
+	return C.jv_string_sized(cstr, C.int(len(str)))
+}
+
+func jvNumber(i interface{}) C.jv {
+	val := reflect.ValueOf(i)
+	switch val.Kind() {
+	case reflect.Float32, reflect.Float64:
+		return C.jv_number(C.double(val.Float()))
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return C.jv_number(C.double(float64(val.Int())))
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return C.jv_number(C.double(float64(val.Uint())))
+	}
+
+	panic("unknown type for go number")
+}
+
+var (
+	// libjq uses callbacks for error handling.
+	// This map safely stores errors under a key for a particular call.
+	// See https://github.com/golang/go/wiki/cgo#function-variables
+	callbackErrors  = make(map[uint64]chan error)
```

Comment:
Why not just plumb this through to each func instead of having a mutex guarded global?

Reply:
How will the callback get it? :|

Reply:
Oh i see you've got some C in here. Can you not pass it into `register_jq_error_cb`?